File: native/cpp/tryops_policy/tests/test_policy.cpp
#include "tryops_policy.hpp"

#include <cassert>
#include <iostream>
#include <string>

namespace {

tryops::Candidate good_vton_candidate() {
  tryops::Candidate candidate;
  candidate.candidate_id = "vton-catvton-2026-06-11-001";
  candidate.workload = "vton";
  candidate.model_name = "catvton-baseline";
  candidate.model_version = "0.1.0";
  candidate.metrics = {
      {"garment_fidelity", 0.81},
      {"identity_preservation", 0.78},
      {"artifact_rate", 0.08},
      {"latency_p95_ms", 9300.0},
  };
  candidate.artifacts = {
      {"model_card", "s3://tryops-artifacts/model-cards/vton-catvton-001.md"},
      {"data_card", "s3://tryops-artifacts/data-cards/vitonhd-demo-v1.md"},
      {"evaluation_report", "s3://tryops-artifacts/reports/vton-catvton-001.json"},
      {"sbom", "s3://tryops-artifacts/sbom/vton-catvton-001.spdx.json"},
      {"model_artifact_scan", "s3://tryops-artifacts/model-scans/vton-catvton-001.json"},
      {"model_provenance", "s3://tryops-artifacts/provenance/vton-catvton-001.model-provenance.json"},
  };
  candidate.metadata = {
      {"model_provenance.status", "passed"},
      {"model_provenance.statement_type", "https://in-toto.io/Statement/v1"},
      {"model_provenance.predicate_type", "https://slsa.dev/provenance/v1"},
      {"model_provenance.signature_mode", "local-dsse-digest"},
      {"model_provenance.signer_identity", "tryops-local-ci"},
      {"model_provenance.verified", "true"},
      {"model_artifacts.serialization_policy", "safetensors_only"},
      {"model_artifacts.scan_status", "passed"},
      {"model_artifacts.unsafe_file_count", "0"},
      {"model_artifacts.safetensors_files", "1"},
      {"model_artifacts.rejected_extensions", ""},
  };
  candidate.approvals = {"mlops_owner", "risk_owner"};
  candidate.risk_status = "medium_approved";
  candidate.critical_vulnerabilities = 0;
  candidate.high_vulnerabilities = 0;
  candidate.signed_artifact = true;
  return candidate;
}

tryops::Candidate bad_vton_candidate() {
  tryops::Candidate candidate = good_vton_candidate();
  candidate.candidate_id = "vton-catvton-2026-06-11-bad";
  candidate.metrics["garment_fidelity"] = 0.61;
  candidate.metrics["identity_preservation"] = 0.65;
  candidate.metrics["artifact_rate"] = 0.21;
  candidate.metrics["latency_p95_ms"] = 18100.0;
  candidate.artifacts.erase("data_card");
  candidate.artifacts.erase("sbom");
  candidate.approvals = {"mlops_owner"};
  candidate.risk_status = "unreviewed";
  candidate.critical_vulnerabilities = 1;
  candidate.high_vulnerabilities = 2;
  candidate.signed_artifact = false;
  return candidate;
}

bool contains_reason(const tryops::PromotionDecision& decision, const std::string& needle) {
  for (const auto& reason : decision.reasons) {
    if (reason.find(needle) != std::string::npos) {
      return true;
    }
  }
  return false;
}

}  // namespace

int main() {
  const auto good = tryops::evaluate_promotion(good_vton_candidate(), "champion");
  assert(good.approved);
  assert(good.reasons.size() == 1);
  assert(good.reasons[0] == "all promotion gates passed");

  const auto bad = tryops::evaluate_promotion(bad_vton_candidate(), "champion");
  assert(!bad.approved);
  assert(contains_reason(bad, "missing required artifacts"));
  assert(contains_reason(bad, "missing approvals"));
  assert(contains_reason(bad, "unaccepted risk status"));
  assert(contains_reason(bad, "critical vulnerabilities"));
  assert(contains_reason(bad, "candidate artifact is not signed"));
  assert(contains_reason(bad, "garment_fidelity"));

  const auto unknown = tryops::evaluate_promotion(good_vton_candidate(), "production");
  assert(!unknown.approved);
  assert(contains_reason(unknown, "unknown target stage"));

  std::cout << "native tryops_policy tests passed\n";
  return 0;
}

File: tests/test_policy.py
from __future__ import annotations

import json
import sys
import unittest
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
SRC = ROOT / "src"
if str(SRC) not in sys.path:
    sys.path.insert(0, str(SRC))

from tryops.contracts import ModelCandidate
from tryops.policy import evaluate_promotion


class PolicyTests(unittest.TestCase):
    def test_good_vton_candidate_passes_champion_gate(self) -> None:
        payload = json.loads((ROOT / "samples/candidates/vton_candidate_good.json").read_text())
        decision = evaluate_promotion(ModelCandidate.from_dict(payload), target_stage="champion")
        self.assertTrue(decision.approved)
        self.assertEqual(decision.reasons, ["all promotion gates passed"])

    def test_bad_vton_candidate_fails_with_multiple_reasons(self) -> None:
        payload = json.loads((ROOT / "samples/candidates/vton_candidate_bad.json").read_text())
        decision = evaluate_promotion(ModelCandidate.from_dict(payload), target_stage="champion")
        self.assertFalse(decision.approved)
        joined = " ".join(decision.reasons)
        self.assertIn("missing required artifacts", joined)
        self.assertIn("missing approvals", joined)
        self.assertIn("unaccepted risk status", joined)
        self.assertIn("critical vulnerabilities", joined)
        self.assertIn("candidate artifact is not signed", joined)
        self.assertIn("garment_fidelity", joined)

    def test_llm_candidate_requires_passing_guardrail_report(self) -> None:
        payload = _llm_candidate()
        decision = evaluate_promotion(ModelCandidate.from_dict(payload), target_stage="champion")

        self.assertTrue(decision.approved)

    def test_llm_candidate_fails_on_guardrail_leak_verdict(self) -> None:
        payload = _llm_candidate()
        payload["metadata"]["guardrails"] = {
            "status": "blocked",
            "failed_cases": 1,
            "blocked_risk_ids": ["LLM07:2025"],
        }
        decision = evaluate_promotion(ModelCandidate.from_dict(payload), target_stage="champion")

        self.assertFalse(decision.approved)
        joined = " ".join(decision.reasons)
        self.assertIn("unaccepted guardrail verdict", joined)
        self.assertIn("guardrail failed cases", joined)
        self.assertIn("LLM07:2025", joined)

    def test_llm_candidate_fails_on_missing_model_provenance(self) -> None:
        payload = _llm_candidate()
        del payload["artifacts"]["model_provenance"]
        payload["metadata"]["model_provenance"] = {
            "status": "missing",
            "statement_type": "missing",
            "predicate_type": "missing",
            "signature_mode": "missing",
            "verified": False,
        }
        decision = evaluate_promotion(ModelCandidate.from_dict(payload), target_stage="champion")

        self.assertFalse(decision.approved)
        joined = " ".join(decision.reasons)
        self.assertIn("missing required artifacts", joined)
        self.assertIn("model provenance status", joined)
        self.assertIn("model provenance verification did not pass", joined)

def _llm_candidate() -> dict[str, object]:
    return {
        "candidate_id": "llm-qwen-2026-06-11-001",
        "workload": "llm",
        "model_name": "Qwen2.5-0.5B-Instruct",
        "model_version": "2026-06-11",
        "metrics": {
            "quality_score": 0.91,
            "tokens_per_second": 90.0,
            "latency_p95_ms": 120.0,
            "memory_gb": 1.5,
        },
        "artifacts": {
            "model_card": "reports/generated/llm/model_card.md",
            "data_card": "reports/generated/llm/data_card.md",
            "evaluation_report": "artifacts/eval/llm_baseline/benchmark.json",
            "sbom": "artifacts/eval/supply_chain/sbom.spdx.json",
            "guardrail_report": "artifacts/eval/guardrails/guardrail_report.json",
            "model_artifact_scan": "artifacts/eval/model_supply_chain/safe_model_artifact_scan.json",
            "model_provenance": "artifacts/eval/model_supply_chain/model_provenance.json",
        },
        "approvals": ["mlops_owner", "risk_owner"],
        "risk_status": "low",
        "vulnerabilities": {"critical": 0, "high": 0},
        "signed": True,
        "metadata": {
            "code_version": "local",
            "dataset_version": "golden-prompts-v1",
            "pipeline_run_id": "run-llm-guardrails",
            "guardrails": {
                "status": "passed",
                "failed_cases": 0,
                "blocked_risk_ids": [],
            },
            "model_provenance": {
                "status": "passed",
                "statement_type": "https://in-toto.io/Statement/v1",
                "predicate_type": "https://slsa.dev/provenance/v1",
                "signature_mode": "local-dsse-digest",
                "signer_identity": "tryops-local-ci",
                "verified": True,
            },
            "model_artifacts": {
                "serialization_policy": "safetensors_only",
                "scan_status": "passed",
                "unsafe_file_count": 0,
                "safetensors_files": 1,
                "rejected_extensions": [],
            },
        },
    }


if __name__ == "__main__":
    unittest.main()
